### PR TITLE
[#970] Guide Users to Fine-Tune with Their Own Data

### PR DESCRIPTION
**Pull request by jeff52415**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [x] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.
- This PR addresses issue #970.

#### Changelog
What are the changes made in this PR?
- Added a new feature that allows users to fine-tune their own dataset by running `tune run lora_finetune_single_device --config llama3/8B_qlora_single_device_custom_instruct`.
- Updated the documentation to include instructions for placing a custom dataset in `./train.json` and the required format for `train.json`.

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help.)

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add unit tests for any new functionality
- [x] update docstrings for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [x] run recipe tests via `pytest tests -m integration_test`
- [x] manually run any new or modified recipes with sufficient proof of correctness
	- [x] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)
  
**Commands and Artifacts:**
- Ran the command `tune run lora_finetune_single_device --config llama3/8B_qlora_single_device_custom_instruct` with a custom `train.json`.

### CODE REVIEW

**Comment by RdoubleA** [comment]
Hi @jeff52415, thanks for opening this PR. I've just updated our dataset tutorials with guidance on how to fine tune on your own data and was planning on sharing more widely very soon. Could you take a look and let me know if this is what you need? https://pytorch.org/torchtune/main/tutorials/datasets.html 

The purpose of the dataset classes and all the builders is so users can quickly bootstrap their own data. If it's still not clear how to do this with the above tutorial, please let me know.

**Comment by ebsmothers** [comment]
Hi @jeff52415 thanks for taking the time to open this PR. To elaborate on what @RdoubleA said, the purpose of our datasets code is to provide fundamental building blocks, along with a few canonical open-source datasets that can be built directly and are widely-used. The building blocks should be sufficiently general that it's pretty easy to extend them. E.g. in this particular case I think it should boil down to a config change looking something like this:

```
dataset:
  _component_: torchtune.datasets.instruct_dataset
  train_on_input: True
  source: json
  dataset_path: /tmp/train.json
```

Anything beyond fundamental building blocks (like how to build a dataset from your own custom dataset file) *should* be addressed in our documentation (but if it's not done so sufficiently we should fix it!). What we don't want is "code-as-documentation". The reason for this is that there are N different dimensions that we can provide customization (dataset, model, tokenization, distributed, optimizer, ...) and providing explicit examples for each of these customizations inside our core code will ultimately do more harm than good when it comes to discoverability of our various offerings. Let me know if this makes sense, thanks!

**Comment by jeff52415** [comment]
@ebsmothers Thanks for the reply, it make sense.

**Comment by jeff52415** [comment]
@ebsmothers Thanks for the reply, it make sense.
